File: Caronae/Ride/RideCell.m
#import "UIImageView+crn_setImageWithURL.h"
#import "Ride.h"
#import "RideCell.h"

@implementation RideCell

static NSDateFormatter *dateFormatter;

- (instancetype)initWithCoder:(NSCoder *)aDecoder {
    self = [super initWithCoder:aDecoder];
    if (self) {
        dateFormatter = [[NSDateFormatter alloc] init];
        dateFormatter.dateFormat = @"HH:mm | E | dd/MM";
    }
    return self;
}

- (void)configureCellWithRide:(Ride *)ride {
    [self configureBasicCellWithRide:ride];
    self.accessoryType = UITableViewCellAccessoryDisclosureIndicator;

    if (ride.going) {
        _arrivalDateTimeLabel.text = [NSString stringWithFormat:@"Chegando às %@", self.dateString];
    }
    else {
        _arrivalDateTimeLabel.text = [NSString stringWithFormat:@"Saindo às %@", self.dateString];
    }
}

- (void)configureHistoryCellWithRide:(Ride *)ride {
    [self configureBasicCellWithRide:ride];
    _arrivalDateTimeLabel.text = [NSString stringWithFormat:@"Chegou às %@", self.dateString];
    self.accessoryType = UITableViewCellAccessoryNone;
}

- (void)configureBasicCellWithRide:(Ride *)ride {
    _ride = ride;
    _titleLabel.text = [_ride.title uppercaseString];
    
    // Display first and last names only
    NSArray<NSString *> *names = [_ride.driver.name componentsSeparatedByString:@" "];
    NSString *displayName = names.count > 1 ? [NSString stringWithFormat:@"%@ %@", names.firstObject, names.lastObject] : _ride.driver.name;
    _driverNameLabel.text = displayName;
    
    [self updatePhoto];
    self.color = [CaronaeDefaults colorForZone:_ride.zone];
    
    _badgeLabel.hidden = YES;
}

- (NSString *)dateString {
    return [dateFormatter stringFromDate:_ride.date].capitalizedString;
}

- (void)updatePhoto {
    if (_ride.driver.profilePictureURL.length > 0) {
        [_photo crn_setImageWithURL:[NSURL URLWithString:_ride.driver.profilePictureURL]];
    }
    else {
        _photo.image = [UIImage imageNamed:@"Profile Picture"];
    }
}

- (void)setColor:(UIColor *)color {
    _color = color;
    _titleLabel.textColor = color;
    _arrivalDateTimeLabel.textColor = color;
    _driverNameLabel.textColor = color;
    _photo.layer.borderColor = color.CGColor;
    self.tintColor = color;
}

- (void)setBadgeCount:(int)badgeCount {
    _badgeCount = badgeCount;
    if (badgeCount > 0) {
        _badgeLabel.text = [NSString stringWithFormat:@"%d", badgeCount];
        _badgeLabel.hidden = NO;
    }
    else {
        _badgeLabel.hidden = YES;
    }
}

- (void)setHighlighted:(BOOL)highlighted animated:(BOOL)animated {
    UIColor *badgeBackgroundColor = _badgeLabel.backgroundColor;
    [super setHighlighted:highlighted animated:animated];
    
    if (highlighted) {
        _badgeLabel.backgroundColor = badgeBackgroundColor;
    }
}

@end

File: Caronae/Ride/RideCell.swift
import UIKit

class RideCell: UITableViewCell {
    
    @IBOutlet weak var titleLabel: UILabel!
    @IBOutlet weak var arrivalDateTimeLabel: UILabel!
    @IBOutlet weak var driverNameLabel: UILabel!
    @IBOutlet weak var photo: UIImageView!
    @IBOutlet weak var badgeLabel: UILabel!
    
    var ride: Ride!
    var color: UIColor! {
        didSet {
            titleLabel.textColor = color
            arrivalDateTimeLabel.textColor = color
            driverNameLabel.textColor = color
            photo.layer.borderColor = color.cgColor
            tintColor = color
        }
    }
    
    var badgeCount: Int! {
        didSet {
            if badgeCount > 0 {
                badgeLabel.text = String(badgeCount)
                badgeLabel.isHidden = false
            } else {
                badgeLabel.isHidden = true
            }
        }
    }
    
    let dateFormatter = DateFormatter()
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        
        dateFormatter.dateFormat = "HH:mm | E | dd/MM"
    }
    
    /// Configures the cell with a Ride object, updating the cell's labels and style accordingly.
    ///
    /// - parameter ride: A Ride object.
    func configureCell(with ride: Ride) {
        configureBasicCell(with: ride)
        accessoryType = .disclosureIndicator
        if ride.going {
            arrivalDateTimeLabel.text = String(format: "Chegando às %@", dateString())
        } else {
            arrivalDateTimeLabel.text = String(format: "Saindo às %@", dateString())
        }
    }
    
    /// Configures the cell with a Ride object which belongs to a user's ride history, updating the cell's labels and style accordingly.
    ///
    /// - parameter ride: A Ride object.
    func configureHistoryCell(with ride: Ride) {
        configureBasicCell(with: ride)
        accessoryType = .none
        if ride.going {
            arrivalDateTimeLabel.text = String(format: "Chegou às %@", dateString())
        } else {
            arrivalDateTimeLabel.text = String(format: "Saiu às %@", dateString())
        }
    }
    
    func configureBasicCell(with ride: Ride) {
        self.ride = ride
        titleLabel.text = ride.title.uppercased()
        driverNameLabel.text = ride.driver.shortName
        
        updatePhoto()
        color = PlaceService.instance.color(forZone: ride.region)
        
        badgeLabel.isHidden = true
    }
    
    func dateString() -> String {
        return dateFormatter.string(from: ride.date).capitalized(after: "|")
    }
    
    func updatePhoto() {
        if let profilePictureURL = ride.driver.profilePictureURL, !profilePictureURL.isEmpty {
            photo.crn_setImage(with: URL(string: profilePictureURL))
        } else {
            photo.image = UIImage(named: CaronaePlaceholderProfileImage)
        }
    }
}
